conslogging/bundle.go

```diff
@@ -0,0 +1,266 @@
+func (bb *BundleBuilder) PrefixWriter(prefix string) io.Writer {
+	if builder, ok := bb.logsForTarget[prefix]; ok {
+		return builder
+	}
+
+	writer := &targetLogger{
+		writer:  &strings.Builder{},
+		status:  StatusWaiting,
+		result:  ResultCancelled,
+		started: time.Now(),
+		prefix:  prefix,
+	}
+	bb.logsForTarget[prefix] = writer
+	return writer
+}
+
+// WriteToDisk aggregates all the data in the numerous prefix writers, and generates an Earthly log bundle.
+// These bundles include a manifest generated from the aggregation of the prefixes (targets).
+func (bb *BundleBuilder) WriteToDisk() (string, error) {
+	// Build file and io.Writer for saving log data
+	file, err := os.CreateTemp("", "earthly-log*.tar.gz")
+	if err != nil {
+		return "", errors.Wrapf(err, "could not create tarball")
+	}
+	defer file.Close()
+	bb.cleanup.Add(func() error {
+		return os.Remove(file.Name())
+	})
+
+	gzipWriter := gzip.NewWriter(file)
+	defer gzipWriter.Close()
+
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer tarWriter.Close()
+
+	// Convert targets to manifest representations, get tar headers for data
+	targetData := make([]TargetManifest, 0)
+	for _, lines := range bb.logsForTarget {
+		mt, err := lines.toManifestTarget()
+		if err != nil {
+			// Something was wrong with this targets logs (0 length, or blacklisted name...). Ignore it.
+			continue
+		}
+
+		targetData = append(targetData, mt)
+
+		trimmed := strings.TrimSpace(lines.prefix)
+		escaped := url.PathEscape(trimmed)
+
+		tarWriter.WriteHeader(&tar.Header{
+			Name:       fmt.Sprintf("target/%s", escaped),
+			Size:       int64(lines.writer.Len()),
+			Mode:       0600,
+			ChangeTime: time.Now(),
+		})
+		tarWriter.Write([]byte(lines.writer.String()))
+	}
+
+	// build manifest and permissions
+	mani := bb.buildManifest(targetData)
+	manifestJSON, _ := json.Marshal(mani)
+	tarWriter.WriteHeader(&tar.Header{
+		Name:       "manifest",
+		Size:       int64(len(manifestJSON)),
+		Mode:       0600,
+		ChangeTime: time.Now(),
+	})
+	tarWriter.Write(manifestJSON)
+
+	perm := bb.buildPermissions()
+	permissionsJSON, _ := json.Marshal(perm)
+	tarWriter.WriteHeader(&tar.Header{
+		Name:       "permissions",
+		Size:       int64(len(permissionsJSON)),
+		Mode:       0600,
+		ChangeTime: time.Now(),
+	})
+	tarWriter.Write(permissionsJSON)
+
+	return file.Name(), nil
+}
+
+func (bb *BundleBuilder) buildManifest(targetManifests []TargetManifest) *Manifest {
+	manifest := &Manifest{
+		Version:    1,
+		Duration:   int(time.Since(bb.started).Seconds()),
+		Status:     "complete",
+		Result:     "success",
+		CreatedAt:  time.Now().In(time.UTC),
+		Entrypoint: bb.entrypoint,
+		Targets:    targetManifests,
+	}
+
+	for _, tm := range targetManifests {
+		if tm.Result != ResultSuccess {
+			manifest.Result = tm.Result
+		}
+
+		if tm.Status != StatusComplete {
+			manifest.Status = tm.Status
+		}
+	}
+
+	return manifest
+}
+
+func (bb *BundleBuilder) buildPermissions() *Permissions {
```

Comment:
something feels odd with this function since it returns an near-empty struct. Is it possible to omit this entirely?

Reply:
Yeah, I agree, somewhat. Its there for completeness, since we do have to build a permissions struct. This makes it fit the pattern of the other parts; waiting until we are ready to fully implement permissions.

Reply:
Inlining it, in context, would look like this:

Reply:
How does the server interpret a `permissions` file containing `{"Version": 1, "Users": [], "Orgs": []}`?

It seems like a potential security issue down the road if we default to an empty list meaning "publicly-for-anyone-with-log-link". Maybe we should have an explicit field stating this is the intended behaviour.

Either that or just omit including a permission file all together if we're not able to handle permissions yet.

Reply:
Right now, the server only requires that the file exist, and it doesn't attempt to interpret the file too much. Logs age out, and we are able to migrate them behind the scenes if needed... so I am not too worried.

Reply:
Our user IDs have been switched to UUIDs this morning (still waiting on Orgs), adding a `*` seems like a good thing to do now, especially if it currently ignores it.

Reply:
Yeah, I was thinking the same thing as I typed that.